qiime2/core/testing/tests/test_mapped_actions.py

```diff
@@ -0,0 +1,190 @@
+
+
+class TestCombinatoricallyMappedMethod(ActionTester):
+    ACTION = 'combinatorically_mapped_method'
+
+    def test_match_foo(self):
+        a = Artifact.import_data('C1[Foo]', 'element 1', view_type=str)
+        b = Artifact.import_data('C3[Foo, Foo, Foo]',
+                                 'element 2', view_type=str)
+
+        x, y = self.run_action(a=a, b=b)
+
+        self.assertEqual(repr(x.type), 'C2[Bar, Bar]')
+        self.assertEqual(repr(y.type), 'Foo')
+
+    def test_match_bar_foo(self):
+        a = Artifact.import_data('C1[Bar]', 'element 1', view_type=str)
+        b = Artifact.import_data('C3[Foo, Foo, Foo]',
+                                 'element 2', view_type=str)
+
+        x, y = self.run_action(a=a, b=b)
+
+        self.assertEqual(repr(x.type), 'C2[Baz, Baz]')
+        self.assertEqual(repr(y.type), 'Foo')
+
+    def test_match_baz_misc(self):
+        a = Artifact.import_data('C1[Baz]', 'element 1', view_type=str)
+        b = Artifact.import_data('C3[Foo, Bar, Baz]',
+                                 'element 2', view_type=str)
+
+        x, y = self.run_action(a=a, b=b)
+
+        self.assertEqual(repr(x.type), 'C2[Foo, Foo]')
+        self.assertEqual(repr(y.type), 'Baz')
+
+    def test_mismatch(self):
+        a = Artifact.import_data('Bar', 'element 1', view_type=str)
+        b = Artifact.import_data('C3[Foo, Foo, Foo]',
+                                 'element 2', view_type=str)
+
+        with self.assertRaises(TypeError):
+            self.run_action(a=a, b=b)
+
+
+class TestDoubleBoundVariableMethod(ActionTester):
+    ACTION = 'double_bound_variable_method'
+
+    def test_predicate_on_second(self):
+        a = Artifact.import_data('Bar', 'element 1', view_type=str)
+        b = Artifact.import_data('Bar % Properties("A")',
+                                 'element 2', view_type=str)
+        extra = Artifact.import_data('Foo', 'always foo', view_type=str)
+
+        x, = self.run_action(a=a, b=b, extra=extra)
+
+        self.assertEqual(repr(x.type), 'Baz')
+
+    def test_mismatch(self):
+        a = Artifact.import_data('Foo', 'element 1', view_type=str)
+        b = Artifact.import_data('Bar', 'element 2', view_type=str)
+        extra = Artifact.import_data('Foo', 'always foo', view_type=str)
+
+        with self.assertRaisesRegex(ValueError, 'match.*same output'):
+            self.run_action(a=a, b=b, extra=extra)
+
+
+class TestBoolFlagSwapsOutputMethod(ActionTester):
+    ACTION = 'bool_flag_swaps_output_method'
+
+    def test_true(self):
+        a = Artifact.import_data('Bar', 'element', view_type=str)
+
+        x, = self.run_action(a=a, b=True)
+
+        self.assertEqual(repr(x.type), 'C1[Foo]')
+
+    def test_false(self):
+        a = Artifact.import_data('Bar', 'element', view_type=str)
+
+        x, = self.run_action(a=a, b=False)
+
+        self.assertEqual(repr(x.type), 'Foo')
+
+
+class TestPredicatesPreservedMethod(ActionTester):
```

Comment:
Not related to this test, specifically, but, while playing with this action, I provided a `Foo`, and saw this error:

```
TypeError: Parameter 'a' received an argument of type Foo. An argument of subtype Foo % (Properties('A', 'X', 'Y', 'C', 'B')¹ | Properties('A', 'C', 'B')² | Properties('A', 'X', 'Y', 'B')³ | Properties('A', 'B')⁴ | Properties('A', 'X', 'Y', 'C')⁵ | Properties('B', 'X', 'Y', 'C')⁶ | Properties('C', 'X', 'Y')⁷ | Properties('A', 'C')⁸ | Properties('B', 'C')⁹ | Properties('C')¹⁰ | Properties('A', 'X', 'Y')¹¹ | Properties('B', 'X', 'Y')¹² | Properties('X', 'Y')¹³ | Properties('A')¹⁴ | Properties('B')¹⁵) is required.
```

Unfortunately, it rendered like this in my terminal:

![image](https://user-images.githubusercontent.com/274668/55759081-66f64100-5a0d-11e9-92af-531112ae4567.png)

Just wanted to share (at first I thought I had something like `Properties('C')^1^0` on my hands and was like "uh, what?")